File: productComputerC/ProductComputer.cpp
/*
 * ProductComputer.cpp
 *
 *  Created on: 22.11.2011
 *      Author: christian
 */

//#define TIMING

#include "ProductComputer.h"

ProductComputer::ProductComputer() {
}

ProductComputer::~ProductComputer() {
}

#include "BladeArrayRoutines.h"

#include <iostream>

    /**
     * Do initializations
     * @param base The ZeroInf Base to use
     * @param base2 The PlusMinus Base to use (here are the products easily defined)
     * @param mapToPlusMinus The map from ZeroInf base to PlusMinus base
     * @param mapToZeroInf The map from PlusMinus base to ZeroInf base
     */
	void ProductComputer::initialize(const Algebra& algebra) {
		//create blades in "blades"
		intVector intBase;
		intBase.reserve(algebra.getBase().size());
		BaseVectors baseVectors;
		string one = "1";
		baseVectors.addBase(one);
		for (unsigned int i=0;i<algebra.getBase().size();i++)
			intBase.push_back(baseVectors.addBase(algebra.getBase()[i]));

		BladeArrayRoutines bladeArrayRoutines;

		vector<Blade> bladesZI;
		bladeArrayRoutines.createBlades(intBase,bladesZI);

		listOfBlades.setBlades(bladesZI);
		//add base2
		for (unsigned int i=0;i<algebra.getBase2().size();i++)
			if (!baseVectors.containsBase(algebra.getBase2()[i]))
				baseVectors.addBase(algebra.getBase2()[i]);

		//baseSquares to integer system

		BaseSquareMap map = algebra.getBaseSquaresStr();
		for (BaseSquareMap::const_iterator ci = map.begin(); ci != map.end(); ++ci) {
			const std::pair<string,int> pair = *ci;
			baseSquares[baseVectors.getIndex(pair.first)] = pair.second;
		}

		//base change of blades to plus minus
		mapStringBladestrlist mapPlusminusStr = algebra.getMapToPlusMinus();
		mapIntBladearray mapBaseChangeToPlusMinus;
		convertMap(mapPlusminusStr, baseVectors, mapBaseChangeToPlusMinus);

		mapStringBladestrlist mapZeroInfStr = algebra.getMapToZeroInf();
		convertMap(mapZeroInfStr, baseVectors, mapBaseChangeToZeroInf);

		for (unsigned int i=0;i<bladesZI.size();i++) {
			SumOfBlades s;
			changeBaseOfBlade(bladesZI[i], mapBaseChangeToPlusMinus, s);
			bladesPM.push_back(s);
		}

		//checking if some blades are zero
		for (unsigned int i=0;i<bladesZI.size();i++) {
			bladesPM[i].checkIfSomeBladesAreZero();
			bladesPM[i].normalize();
			bladesPM[i].group();
		}

	}

	typedef list<Blade> listBlade;

#ifdef TIMING
#include "Timing.h"
#endif


	void ProductComputer::calcProduct(const int i, const int j, ProductCalculator& calculator, Multivector& result) {
        //use bladesPM and mapBaseChangeToZeroInf
#ifdef TIMING
		Timing::timing.startTime("1");
#endif

        //calculate real product
        list<Blade>& blades1 = bladesPM[i].blades;
        list<Blade>& blades2 = bladesPM[j].blades;
        SumOfBlades product;


        const listBlade::const_iterator& endItBlades1 = blades1.end();
        const listBlade::const_iterator& endItBlades2 = blades2.end();
        for (listBlade::const_iterator ci1 = blades1.begin(); ci1 != endItBlades1;++ci1)
            for (listBlade::const_iterator ci2 = blades2.begin(); ci2 != endItBlades2;++ci2) {
            	calculator.calculate(*ci1, *ci2, baseSquares, product);
            }
        product.checkIfSomeBladesAreZero();
#ifdef TIMING
        Timing::timing.stopTime("1");
        Timing::timing.startTime("2");
#endif

        //base change
        SumOfBlades blade;
        changeBaseOfBlade(product, mapBaseChangeToZeroInf, blade);
        //12s
#ifdef TIMING
        Timing::timing.stopTime("2");
        Timing::timing.startTime("3");
#endif
        //checking if some blades are zero
        blade.checkIfSomeBladesAreZero();
#ifdef TIMING
        Timing::timing.stopTime("3");
        Timing::timing.startTime("4");
#endif

        //normalise blades
        blade.normalize();
#ifdef TIMING
        Timing::timing.stopTime("4");
        Timing::timing.startTime("5");
#endif
         //group blades
        blade.group();
#ifdef TIMING
        Timing::timing.stopTime("5");
        Timing::timing.startTime("6");
#endif
        //merge to multivector
        blade.toMultivector(listOfBlades, result);
#ifdef TIMING
        Timing::timing.stopTime("6");
#endif

    }

    /**
     * Converts a map from a user-friendly format in a efficient format for accesses
     * @param mapBaseChangeStr The map to convert
     * @param baseVectors The initialized BaseVectors object for finding indices
     * @return The new map
     */
    void ProductComputer::convertMap(mapStringBladestrlist& mapBaseChangeStr,  BaseVectors& baseVectors, mapIntBladearray& result) {
		for (mapStringBladestrlist::const_iterator ci = mapBaseChangeStr.begin(); ci != mapBaseChangeStr.end(); ++ci) {
			const std::pair<string, bladeStrList> pair = *ci;

			BladeArray bladeArray;
            vector<Blade>& blades = bladeArray.blades;
            for (bladeStrList::const_iterator ci1 = pair.second.begin(); ci1 != pair.second.end(); ++ci1) {
            	BladeStr bladeStr = *ci1;
            	Blade b;
            	convertBladeStrToBlade(bladeStr, baseVectors, b);
            	blades.push_back(b);
            }

            result[baseVectors.getIndex(pair.first)] = bladeArray;
        }
    }

    /**
     * Converts a BladeStr object to a Blade object
     * @param bladeStr The BladeStr object
     * @param baseVectors The baseVectors for finding indices
     * @return The new Blade object
     */
    void ProductComputer::convertBladeStrToBlade(const BladeStr& bladeStr,  BaseVectors& baseVectors, Blade& result) {
        const vector<string>& baseVectorsStr = bladeStr.baseVectors;

        intVector& baseV = result.baseVectors;
        baseV.clear();
        baseV.reserve(baseVectorsStr.size());
        for (unsigned int i=0;i<baseVectorsStr.size();i++)
            baseV.push_back(baseVectors.getIndex(baseVectorsStr[i]));

        result.setPrefactor(bladeStr.getPrefactor());
    }

    /**
     * Changes the base for a sum of blades and saves the result in a SumOfBlades object
     * e1^einf^e3 -> e1^em^e3 + e1^ep^e3
     * @param sumOfBlades The sum of blades
     * @param map The map to use for the basechange
     * @return The new SumOfBlades object
     */
    void ProductComputer::changeBaseOfBlade(const SumOfBlades& sumOfBlades, unordered_map<int, BladeArray>& map, SumOfBlades& resultB) {
    	const listBlade& blades = sumOfBlades.blades;
    	const listBlade::const_iterator& endIt = blades.end();
        for (listBlade::const_iterator ci=blades.begin(); ci != endIt; ++ci)
        	changeBaseOfBlade(*ci, map, resultB);
    }

#include "PermutableIterator.h"

    /**
     * Changes the base for a blade and saves the result in a SumOfBlades object
     * e1^einf^e3 -> e1^em^e3 + e1^ep^e3
     * @param blade The blade
     * @param map The map to use for the basechange
     * @return The new SumOfBlades object
     */
    void ProductComputer::changeBaseOfBlade(const Blade& blade, unordered_map<int, BladeArray>& map, SumOfBlades& resultB) {
    	const intVector& baseVectors = blade.baseVectors;
        if (baseVectors.size() == 0) {
        	resultB.addBlade(blade);
        	return;
        }

        //find the lengths for creating a permutation
        intVector lengths;

        lengths.reserve(baseVectors.size());
        int j = 0;
        const intVector::const_iterator& endItBaseVectors0 = baseVectors.end();
        for (intVector::const_iterator ci=baseVectors.begin(); ci != endItBaseVectors0; ++ci) {
        	const int i = *ci;
            lengths.push_back((map.count(i) == 1)
                    ? map[i].blades.size()
                    : 1);
            j++;
        }


        //create the permutations
        PermutableIterator iterator;
        intVector permutation;
        iterator.initialize(lengths, permutation);

        do {

            //build current permutation in baseList
        	Blade c;
            intVector& baseList = c.baseVectors;
            baseList.reserve(lengths.size());

            float prefactor = 1;
            j=0;



            const intVector::const_iterator& endItBaseVectors = baseVectors.end();
            for (intVector::const_iterator ci=baseVectors.begin(); ci != endItBaseVectors; ++ci) {
            	const int i = *ci;

            	unordered_map<int, BladeArray>::iterator it = map.find(i);
                if (it != map.end()) {
                    Blade b;
                    (*it).second.getBlade(permutation[j],b);
                    prefactor *= b.getPrefactor();
                    const intVector::const_iterator& endItBBaseVectors = b.baseVectors.end();
                    for (intVector::const_iterator ci2 = b.baseVectors.begin(); ci2 != endItBBaseVectors; ++ci2)
                    	baseList.push_back(*ci2);
                } else {
                    baseList.push_back(i);
                }
                j++;
            }


            c.setPrefactor(prefactor*blade.getPrefactor());
            resultB.addBlade(c);
            iterator.getNextPermutation(permutation);

        } while (iterator.curNo<iterator.count);
    }

File: tba/src/main/java/de/gaalop/productComputer2/ProductComputer.java
package de.gaalop.productComputer2;

import de.gaalop.tba.Multivector;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;

/**
 * This class computes products
 * @author christian
 */
public class ProductComputer {

    private HashMap<Integer, BladeArray> mapBaseChangeToZeroInf;
    private ListOfBlades listOfBlades;
    private SumOfBlades[] bladesPM;
    private HashMap<Integer, Byte> baseSquares;

    /**
     * Do initializations
     * @param base The ZeroInf Base to use
     * @param base2 The PlusMinus Base to use (here are the products easily defined)
     * @param mapToPlusMinus The map from ZeroInf base to PlusMinus base
     * @param mapToZeroInf The map from PlusMinus base to ZeroInf base
     */
    public void initialize(AlgebraPC algebra) {
        //create blades in "blades"
        int[] intBase = new int[algebra.base.length];
        BaseVectors baseVectors = new BaseVectors();
        baseVectors.addBase("1");
        for (int i=0;i<algebra.base.length;i++)
            intBase[i] = baseVectors.addBase(algebra.base[i]);

        Blade[] bladesZI = BladeArrayRoutines.createBlades(intBase);
        listOfBlades = new ListOfBlades(bladesZI);
        //add base2
        for (int i=0;i<algebra.base2.length;i++)
            if (!baseVectors.containsBase(algebra.base2[i]))
                baseVectors.addBase(algebra.base2[i]);

        //baseSquares to integer system
        baseSquares = new HashMap<Integer, Byte>();
        for (String key: algebra.baseSquaresStr.keySet())
            baseSquares.put(baseVectors.getIndex(key), algebra.baseSquaresStr.get(key));

        //base change of blades to plus minus
        HashMap<Integer, BladeArray> mapBaseChangeToPlusMinus = convertMap(algebra.mapToPlusMinus, baseVectors);
        mapBaseChangeToZeroInf = convertMap(algebra.mapToZeroInf, baseVectors);
        bladesPM = new SumOfBlades[bladesZI.length];
        for (int i=0;i<bladesZI.length;i++)
            bladesPM[i] = changeBaseOfBlade(bladesZI[i], mapBaseChangeToPlusMinus);

        //checking if some blades are zero
        for (int i=0;i<bladesZI.length;i++) {
            bladesPM[i].checkIfSomeBladesAreZero();
            bladesPM[i].normalize();
            bladesPM[i].group();
        }

        //test printing
        //for (int i=0;i<blades.length;i++)
        //    System.out.println(blades[i].print(baseVectors)+" = "+bladesPM[i].print(baseVectors));
    }

    public Multivector calcProduct(int i, int j, ProductCalculator calculator) {
        //use bladesPM and mapBaseChangeToZeroInf

        //calculate real product
        SumOfBlades product = new SumOfBlades(new LinkedList<Blade>());
        LinkedList<Blade> blades1 = bladesPM[i].getBlades();
        LinkedList<Blade> blades2 = bladesPM[j].getBlades();
        LinkedList<Blade> prodBlades = product.getBlades();

        for (Blade blade1: blades1)
            for (Blade blade2: blades2)
                prodBlades.addAll(calculator.calculate(blade1, blade2, baseSquares).getBlades());

        product.checkIfSomeBladesAreZero();
        //base change
        SumOfBlades blade = changeBaseOfBlade(product, mapBaseChangeToZeroInf);

        //checking if some blades are zero
        blade.checkIfSomeBladesAreZero();
        //normalise blades
        blade.normalize();
         //group blades
        blade.group();

        //merge to multivector
        return blade.toMultivector(listOfBlades);
    }

    /**
     * Converts a map from a user-friendly format in a efficient format for accesses
     * @param mapBaseChangeStr The map to convert
     * @param baseVectors The initialized BaseVectors object for finding indices
     * @return The new map
     */
    private static HashMap<Integer, BladeArray> convertMap(HashMap<String, LinkedList<BladeStr>> mapBaseChangeStr, BaseVectors baseVectors) {
        HashMap<Integer, BladeArray> mapBaseChange = new HashMap<Integer, BladeArray>();
        for (String key: mapBaseChangeStr.keySet()) {
            LinkedList<BladeStr> listOfStrBlade = mapBaseChangeStr.get(key);
            Blade[] blades = new Blade[listOfStrBlade.size()];
            int i=0;
            for (BladeStr bladeStr: listOfStrBlade)
                blades[i++] = convertBladeStrToBlade(bladeStr, baseVectors);
            mapBaseChange.put(baseVectors.getIndex(key), new BladeArray(blades));
        }
        return mapBaseChange;
    }

    /**
     * Converts a BladeStr object to a Blade object
     * @param bladeStr The BladeStr object
     * @param baseVectors The baseVectors for finding indices
     * @return The new Blade object
     */
    private static Blade convertBladeStrToBlade(BladeStr bladeStr, BaseVectors baseVectors) {
        String[] baseVectorsStr = bladeStr.getBaseVectors();

        LinkedList<Integer> baseV = new LinkedList<Integer>();
        for (int i=0;i<baseVectorsStr.length;i++) 
            baseV.add(baseVectors.getIndex(baseVectorsStr[i]));

        return new Blade(bladeStr.getPrefactor(), baseV);
    }

    /**
     * Changes the base for a sum of blades and saves the result in a SumOfBlades object
     * e1^einf^e3 -> e1^em^e3 + e1^ep^e3
     * @param sumOfBlades The sum of blades
     * @param map The map to use for the basechange
     * @return The new SumOfBlades object
     */
    private static SumOfBlades changeBaseOfBlade(SumOfBlades sumOfBlades, HashMap<Integer, BladeArray> map) {
        LinkedList<Blade> result = new LinkedList<Blade>();
        for (Blade blade: sumOfBlades.getBlades()) 
            result.addAll(changeBaseOfBlade(blade, map).getBlades());
        return new SumOfBlades(result);
    }

    /**
     * Changes the base for a blade and saves the result in a SumOfBlades object
     * e1^einf^e3 -> e1^em^e3 + e1^ep^e3
     * @param blade The blade
     * @param map The map to use for the basechange
     * @return The new SumOfBlades object
     */
    private static SumOfBlades changeBaseOfBlade(Blade blade, HashMap<Integer, BladeArray> map) {
        LinkedList<Integer> baseVectors = blade.getBaseVectors();
        if (baseVectors.size() == 0) return new SumOfBlades(blade);

        //find the lengths for creating a permutation
        int[] lengths = new int[baseVectors.size()];
        int j = 0;
        for (Integer i: baseVectors) {
            lengths[j] = map.containsKey(i)
                    ? map.get(i).getBlades().length
                    : 1;
            j++;
        }


        //create the permutations
        Permutable iterator = new PermutableIterator();
        iterator.initialize(lengths);

        //apply permutations
        LinkedList<Blade> result = new LinkedList<Blade>();

        while (iterator.hasNextPermutation()) {
            IntArray permutation = iterator.getNextPermutation();
            //build current permutation in baseList
            LinkedList<Integer> baseList = new LinkedList<Integer>();

            float prefactor = 1;
            j=0;
            for (Integer i: baseVectors) {
                if (map.containsKey(i)) {
                    Blade b = map.get(i).getBlades()[permutation.getArray()[j]];
                    prefactor *= b.getPrefactor();
                    baseList.addAll(b.getBaseVectors());
                } else {
                    baseList.add(i);
                }
                j++;
            }

            result.add(new Blade(prefactor*blade.getPrefactor(), baseList));
        }
        
        return new SumOfBlades(result);
    }

}
